packages/search-input/src/SearchInput/SearchInput.tsx

```diff
@@ -0,0 +1,433 @@
+      size = Size.Default,
+      disabled,
+      children,
+      state = State.Unset,
+      value: valueProp,
+      onChange: onChangeProp,
+      onSubmit: onSubmitProp,
+      'aria-label': ariaLabel,
+      'aria-labelledby': ariaLabelledBy,
+      ...rest
+    }: SearchInputProps,
+    forwardRef: React.Ref<HTMLInputElement>,
+  ) {
+    const { theme, darkMode } = useDarkMode(darkModeProp);
+    const [isOpen, setOpen] = useState(false);
+    // The index of the currently highlighted result option
+    const [highlightIndex, setHighlightIndex] = useState<number>(0);
+
+    const closeMenu = () => setOpen(false);
+    const openMenu = () => setOpen(true);
+
+    const formRef = useRef<HTMLFormElement>(null);
+    const searchBoxRef = useRef<HTMLDivElement>(null);
+    const menuRef = useRef<HTMLUListElement>(null);
+    const inputRef = useForwardedRef(forwardRef, null);
+    const clearButtonRef = useRef<HTMLButtonElement>(null);
+    const resultRefs = useDynamicRefs<HTMLElement>({ prefix: 'result' });
+    const [focusedElement, trackFocusedElement] = useState<Element>();
+    const highlightedElementRef = resultRefs(`${highlightIndex}`);
+
+    const withTypeAhead = !isUndefined(children);
+
+    const { value, handleChange } = useControlledValue(valueProp, onChangeProp);
+
+    /** Fires a change event to update the input value */
+    const changeInputValue = useCallback(
+      (newVal: string) => {
+        if (inputRef.current) {
+          // We change the element's value
+          // and then make sure that the change event is fired
+          inputRef.current.value = newVal;
+          const nativeChangeEvent = new Event('change', {
+            cancelable: true,
+            bubbles: true,
+          });
+          const syntheticChangeEvent = createSyntheticEvent(
+            nativeChangeEvent,
+            inputRef.current,
+          );
+          handleChange(syntheticChangeEvent);
+        }
+      },
+      [handleChange, inputRef],
+    );
+
+    /**
+     * Helper function that both counts the number of `SearchResult` descendants
+     * and adds the appropriate props to said children
+     */
+    const processChildren = useCallback(() => {
+      // Count results (not just children, since groups are still children)
+      let resultsCount = 0;
+
+      const processChild = (
+        child: React.ReactNode,
+      ): JSX.Element | undefined => {
+        if (isComponentType(child, 'SearchResult')) {
+          resultsCount += 1;
+          const index = resultsCount - 1;
+
+          const textValue = getNodeTextContent(child);
+
+          const onElementClick: MouseEventHandler = e => {
+            child.props.onClick?.(e); // call the child's onClick handler
+
+            changeInputValue(textValue);
+
+            const wasClickedWithMouse = e.detail >= 1;
+
+            if (wasClickedWithMouse && formRef.current && inputRef.current) {
+              // Selecting an option fires the `submit` event
```

Comment:
Why do we do this?

Reply:
Do you mean the submit event or the mouse check?

We need the submit event so that users can listen for `onSubmit` instead of `onClick` for every result

We need to do the check since the `Enter` key already triggers a `click` event on the result and `submit` on the form